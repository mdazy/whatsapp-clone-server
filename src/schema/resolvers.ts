import { DateTimeResolver, URLResolver } from "graphql-scalars";
import { chats, users, messages, chatRooms } from "../db";

const resolvers = {
  Date: DateTimeResolver,
  URL: URLResolver,
  Chat: {
    user(chat: any) {
      return users.filter(u => u.id === chat.userId)[0];
    },
    lastMessage(chat: any) {
      return messages.filter(m => m.id === chat.lastMessageId)[0];
    }
  },
  ChatRoom: {
    messages(chat: any) {
      const chatRoom = chatRooms.filter(c => c.id === chat.id)[0];
      return chatRoom.messageIds.map(
        id => messages.filter(mm => mm.id === id)[0]
      );
    }
  },
  Query: {
    chats() {
      return chats;
    },
    chat(_: any, params: any) {
      const id = Number(params.id);
      let chatRoom = chatRooms.filter(c => c.id === id);
      if (chatRoom.length === 0) {
        // hmmm which do we want?
        // throw new Error(`No chat with id ${id}`);
        return undefined;
      }
      return { id };
    }
  }
};

export default resolvers;
