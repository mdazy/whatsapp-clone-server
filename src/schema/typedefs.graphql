scalar Date
scalar URL

type Message {
  id: ID!
  fromUser: User!
  content: String!
  date: Date!
}

type User {
  id: ID!
  name: String!
  image: URL!
}

type Chat {
  id: ID!
  user: User!
  lastMessage: Message!
  messages: [Message!]!
}

type Query {
  chats: [Chat!]!
  chat(id: ID): Chat
}

type Mutation {
  sendMessage(chatId: ID!, userId: ID!, content: String!): Message
}
