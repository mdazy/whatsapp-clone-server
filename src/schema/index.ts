import { importSchema } from "graphql-import";
import { makeExecutableSchema } from "graphql-tools";
import resolvers from "./resolvers";

const typeDefs = importSchema("src/schema/typedefs.graphql");

export default makeExecutableSchema({ resolvers, typeDefs });
