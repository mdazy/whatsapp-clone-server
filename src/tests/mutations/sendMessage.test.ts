import { createTestClient } from "apollo-server-testing";
import { ApolloServer } from "apollo-server-express";
import schema from "../../schema";

describe("Mutation.sendMessage", () => {
  it("should add a new message", async () => {
    const server = new ApolloServer({ schema });
    const { query, mutate } = createTestClient(server);

    const res = await mutate({
      mutation: `
        mutation {
            sendMessage(chatId: 1, userId:2, content: "pika") {
            id
            content
            }
        }
      `
    });

    expect(res.data).toBeDefined();
    expect(res.data).toMatchSnapshot();
    expect(res.errors).toBeUndefined();
  });

  it("should throw for non-existing chat", async () => {
    const server = new ApolloServer({ schema });
    const { mutate } = createTestClient(server);

    const res = await mutate({
      mutation: `mutation { sendMessage(chatId: 5, userId: 2, content: "pika") { id } }`
    });

    expect(res.data).toBeDefined();
    expect(res.data).toEqual({ sendMessage: null });
    expect(res.errors).toBeDefined();
    expect(res.errors).toMatchSnapshot();
  });

  it("should throw for non-existing user", async () => {
    const server = new ApolloServer({ schema });
    const { mutate } = createTestClient(server);

    const res = await mutate({
      mutation: `mutation { sendMessage(chatId: 1, userId: 5, content: "pika") { id } }`
    });

    expect(res.data).toBeDefined();
    expect(res.data).toEqual({ sendMessage: null });
    expect(res.errors).toBeDefined();
    expect(res.errors).toMatchSnapshot();
  });
});
