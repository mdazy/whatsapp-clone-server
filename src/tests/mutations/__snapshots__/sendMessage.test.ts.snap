// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mutation.sendMessage should add a new message 1`] = `
Object {
  "sendMessage": Object {
    "content": "pika",
    "id": "4",
  },
}
`;

exports[`Mutation.sendMessage should throw for non-existing chat 1`] = `
Array [
  [GraphQLError: No chat with id 5],
]
`;

exports[`Mutation.sendMessage should throw for non-existing user 1`] = `
Array [
  [GraphQLError: No user with id 5],
]
`;
